src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -382,67 +336,55 @@ private void ForceAuthentication(bool receiveFirst, byte[] buffer, AsyncProtocol
-            {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndAuthenticate"));
+                await InnerStream.WriteAsync(message.Payload, cancellationToken).ConfigureAwait(false);
             }
 
-            InternalEndProcessAuthentication(lazyResult);
-
-            // Connection is completed at this point.
-            if (NetEventSource.IsEnabled)
+            do
             {
-                if (NetEventSource.IsEnabled)
-                    NetEventSource.Log.SspiSelectedCipherSuite(nameof(EndProcessAuthentication),
-                                                                SslProtocol,
-                                                                CipherAlgorithm,
-                                                                CipherStrength,
-                                                                HashAlgorithm,
-                                                                HashStrength,
-                                                                KeyExchangeAlgorithm,
-                                                                KeyExchangeStrength);
-            }
-        }
+                message  = await ReceiveBlobAsync(adapter, buffer, cancellationToken).ConfigureAwait(false);
+                if (message.Size > 0)
+                {
+                    // If there is message send it out even if call failed. It may contain TLS Altert.
+                    await InnerStream.WriteAsync(message.Payload, cancellationToken).ConfigureAwait(false);
+                }
 
-        private void InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
-        {
-            // No "artificial" timeouts implemented so far, InnerStream controls that.
-            lazyResult.InternalWaitForCompletion();
-            Exception e = lazyResult.Result as Exception;
+                if (message.Failed)
+                {
+                    throw new AuthenticationException(SR.net_auth_SSPI, message.GetException());
+                }
+            } while (message.Status.ErrorCode != SecurityStatusPalErrorCode.OK);
 
-            if (e != null)
+            ProtocolToken alertToken = null;
+            if (!CompleteHandshake(ref alertToken))
             {
-                // Failed auth, reset the framing if any.
-                _Framing = Framing.Unknown;
-                _handshakeCompleted = false;
-
-                SetException(e);
-                ThrowIfExceptional();
+                StartSendAuthResetSignal(alertToken, ExceptionDispatchInfo.Capture(new AuthenticationException(SR.net_ssl_io_cert_validation, null)));
             }
         }
 
         //
         // Client side starts here, but server also loops through this method.
         //
-        private void StartSendBlob(byte[] incoming, int count, AsyncProtocolRequest asyncRequest)
+        private void StartSendBlob(byte[] incoming, int count)
```

Comment:
I'm confused by this "Start" naming.  Is this still initiating an async operation?  If not, can it just be "SendBlob"?  Same goes for other similar cases.

Reply:
I updated names and collapsed some functions together.